### PR TITLE
to_torchscript method for LightningModule

### LINKED ISSUE #3080: Support a to_torchscript function on the LightningModule
## 🚀 Feature
Support a conversion function to [PyTorch JIT](https://pytorch.org/docs/stable/jit.html) similar to what's available for [ONNX](https://pytorch-lightning.readthedocs.io/en/latest/lightning-module.html#to-onnx).
 
### Motivation
TorchScript is a way to create serializable and optimizable models from PyTorch code. Any TorchScript program can be saved from a Python process and loaded in a process where there is no Python dependency. TorchScript is a method by which users can serve PyTorch models efficiently. 

### Pitch
By default, we can use TorchScript to script the LightningModule. Users can override this in their own lightning modules to use tracing, or to script specific nn.Modules inside their LightningModule. This can then be extended to other Lightning utilities like model checkpointing, so we can save TorchScript or ONNX converted models alongside the best model checkpoints to make going to serving even easier to do

```
    def to_torchscript(self):
        """Saves the model as a JIT module. 
            This can be overridden to support custom TorchScript module export
        Example:
            >>> class SimpleModel(LightningModule):
            ...     def __init__(self):
            ...         super().__init__()
            ...         self.l1 = torch.nn.Linear(in_features=64, out_features=4)
            ...
            ...     def forward(self, x):
            ...         return torch.relu(self.l1(x.view(x.size(0), -1)))
            >>> with tempfile.NamedTemporaryFile(suffix='.onnx', delete=False) as tmpfile:
            ...     model = SimpleModel()
            ...     torch.jit.save(model.to_torchscript(), tmpfile.name)
            ...     os.path.isfile(tmpfile.name)
            True
        """
       return torch.jit.script(self.eval())
```

### PR DESCRIPTION
**Pull request by awaelchli**
# What does this PR do?

Fixes #3080

Tests:
- tests exporting several modules (GAN, RNN, Standard test model)
- test save and load
- test that unused properties like datamodules etc, and hooks get removed from scriptmodule

**NOTE:** 
- Renamed TestGAN because in pytest, classes named Test-... are treated as pytest-TestClasses which are supposed to hold testing methods and this messes with torchscript somehow.
- Had to skip some tests because torch.jit.save crashes on pytorch <= 1.4. I assume it's a bug that got fixed, I don't understand the cause though. 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1729,6 +1730,33 @@ def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwarg
...
+            ...     def forward(self, x):
+            ...         return torch.relu(self.l1(x.view(x.size(0), -1)))
+            ...
+            >>> model = SimpleModel()
+            >>> torch.jit.save(model.to_torchscript(), "model.jit")
+            >>> os.path.isfile("model.jit")
+            True
+        """
+        mode = self.training
+        scripted_module = torch.jit.script(self.eval())
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1756]: Wrap the script call with `torch.no_grad()` ?

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1756]: good idea! currently working on tests.

File: tests/base/models.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def forward(self, img):
 class TestGAN(LightningModule):
     """Implements a basic GAN for the purpose of illustrating multiple optimizers."""
 
-    def __init__(self, hidden_dim, learning_rate, b1, b2, **kwargs):
+    def __init__(self, hidden_dim=128, learning_rate=0.001, b1=0.5, b2=0.999, **kwargs):
```
Comment (ananthsub) [inline tests/base/models.py:70]: these could use typehints. Is there an open GH issue for adding typehints to more of the lightning codebase, or to tests specifically?

Comment (awaelchli) [reply tests/base/models.py]: thanks! we had such an issue, but after adding typehints to the majority of the codebase, we closed it. there are only a few spots like this one where we forgot them. fixed now :)

File: docs/source/production_inference.rst [hunk 3]
```diff
@@ -28,3 +28,21 @@ Once you have the exported model, you can run it on your ONNX runtime in the fol
...
+
+.. code-block:: python
+
+    model = SimpleModel()
+    script = model.to_torchscript()
+
+    # save for use in production environment
+    torch.jit.save(script, "model.pt")
+
+It is recommended that you install the latest version of PyTorch to use this feature without limitations.
```
Comment (justusschock) [inline docs/source/production_inference.rst:48]:
```suggestion
It is recommended that you install the latest supported version of PyTorch to use this feature without limitations.
```

sometimes we're a week or so behind with fully supporting the newest version :)

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1729,6 +1730,36 @@ def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwarg
...
         torch.onnx.export(self, input_data, file_path, **kwargs)
 
+    def to_torchscript(self) -> torch.jit.ScriptModule:
+        """
+        Compiles the model to a :class:`~torch.jit.ScriptModule`.
+        This can be overridden to support custom TorchScript module export.
+
+        Note:
+            - Requires the implementation of the :meth:`LightningModule.forward` method.
+            - It is recommended that you install the latest version of PyTorch to use this feature without limitations.
```
Comment (justusschock) [inline pytorch_lightning/core/lightning.py:1740]:
```suggestion
            - It is recommended that you install the latest supported version of PyTorch to use this feature without limitations.
```

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1729,6 +1730,36 @@ def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwarg
...
+            >>> class SimpleModel(LightningModule):
+            ...     def __init__(self):
+            ...         super().__init__()
+            ...         self.l1 = torch.nn.Linear(in_features=64, out_features=4)
+            ...
+            ...     def forward(self, x):
+            ...         return torch.relu(self.l1(x.view(x.size(0), -1)))
+            ...
+            >>> model = SimpleModel()
+            >>> torch.jit.save(model.to_torchscript(), "model.pt")  # doctest: +SKIP
```
Comment (justusschock) [inline pytorch_lightning/core/lightning.py:1753]: why skipping these?

**Review by justusschock — APPROVED** [review]
LGTM.

I only have one question: Theoretically it is possible to also train ScriptModules. But this PR does not enable this, right? Because that way saving and loading may have to properly deal with the attributes. 

Not sure, what happens if you derive your model from `LightningModule` and `ScriptModule` though

File: pytorch_lightning/core/lightning.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1753]: torchscript needs the source file to work, but in doctests it doesn't seem to be working, errors with something like "can't find source file". so I kept it for documentation, as an example. I know it's a bit dumb... better to remove? @justusschock

Comment (justusschock) [reply pytorch_lightning/core/lightning.py:1755]: No, was just wondering why it did not work. Think they should stay as examples.

File: docs/source/production_inference.rst [hunk 6]
```diff
@@ -28,3 +28,21 @@ Once you have the exported model, you can run it on your ONNX runtime in the fol
...
+
+.. code-block:: python
+
+    model = SimpleModel()
+    script = model.to_torchscript()
+
+    # save for use in production environment
+    torch.jit.save(script, "model.pt")
+
+It is recommended that you install the latest supported version of PyTorch to use this feature without limitations.
```
Comment (SkafteNicki) [inline docs/source/production_inference.rst:48]: Maybe we should insert link to https://pytorch.org/docs/stable/jit.html?

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1729,6 +1730,38 @@ def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwarg
...
+            >>> os.path.isfile("model.pt")  # doctest: +SKIP
+            True
+        """
+        mode = self.training
+        with torch.no_grad():
+            scripted_module = torch.jit.script(self.eval())
```
Comment (SkafteNicki) [inline pytorch_lightning/core/lightning.py:1761]: I think it should be explicit stated that the scripted model is saved in evaluation mode.

**Comment by awaelchli** [comment]
> I only have one question: Theoretically it is possible to also train ScriptModules. But this PR does not enable this, right? Because that way saving and loading may have to properly deal with the attributes.

my understanding is scripting removes all properties, methods etc. that do not appear in forward (I assert that in thet test actually). all the lightning stuff is gone. haven't tried, but if you were to train a script module, it would have to use forward, so pure pytorch training loop. 

we talked with @ananthsub of a way to load a script module into a LightningModule, that would essentially replace the forward and then could be trained. if we can formulate a concrete plan for how to integrate that, I'm happy to follow up on it.

File: docs/source/production_inference.rst [hunk 8]
```diff
@@ -28,3 +28,21 @@ Once you have the exported model, you can run it on your ONNX runtime in the fol
...
+The LightningModule has a handy method :meth:`~pytorch_lightning.core.lightning.LightningModule.to_torchscript`
+that returns a scripted module which you can save or directly use.
+
+.. code-block:: python
+
+    model = SimpleModel()
+    script = model.to_torchscript()
+
+    # save for use in production environment
+    torch.jit.save(script, "model.pt")
```
Comment (rohitgr7) [inline docs/source/production_inference.rst:46]: maybe we should have a `file_path` just like in `to_onnx`?? If it's None, return the script else save it in a file.

Comment (justusschock) [reply docs/source/production_inference.rst:46]: I think it should have a file_path but it should also always return the script (and only optionally save)

File: pytorch_lightning/core/lightning.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/core/lightning.py:1758]: maybe add a comment about the skip to the code

**Review by Borda — APPROVED** [review]

File: docs/source/production_inference.rst (hunk 8 above)
Comment (awaelchli) [reply docs/source/production_inference.rst:46]: @ananthsub thoughts on this?

Comment (ananthsub) [reply docs/source/production_inference.rst:46]: I'm fine with an optional file_path. This would provide some consistency with `to_onnx`. But I agree with @justusschock that it should always return the script module

Comment (awaelchli) [reply docs/source/production_inference.rst:46]: done, documented and tested.

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -1729,6 +1730,54 @@ def to_onnx(self, file_path: str, input_sample: Optional[Tensor] = None, **kwarg
 
         torch.onnx.export(self, input_data, file_path, **kwargs)
 
+    def to_torchscript(self, file_path: str = None, **kwargs) -> Union[ScriptModule, Dict[str, ScriptModule]]:
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1733]:
```suggestion
    def to_torchscript(self, file_path: Optional[str] = None, **kwargs) -> Union[ScriptModule, Dict[str, ScriptModule]]:
```

**Review by rohitgr7 — APPROVED** [review]
NICE!!
